sway/input/input-manager.c

```diff
@@ -507,6 +541,31 @@ static void handle_new_input(struct wl_listener *listener, void *data) {
 	sway_log(SWAY_DEBUG, "adding device: '%s'",
 		input_device->identifier);
 
+	const char *device_type = input_device_get_type(input_device);
+	struct input_config *type_config = NULL;
+	for (int i = 0; i < config->input_type_configs->length; i++) {
+		struct input_config *ic = config->input_type_configs->items[i];
+		if (strcmp(ic->identifier + 5, device_type) == 0) {
+			type_config = ic;
+			break;
+		}
+	}
+
+	// Merge new device config on top of type wildcard, if the prior exists
+	for (int i = 0; i < config->input_configs->length; i++) {
+		struct input_config *ic = config->input_configs->items[i];
+		if (strcmp(input_device->identifier, ic->identifier) == 0) {
+			struct input_config *current = new_input_config(ic->identifier);
+			merge_input_config(current, type_config);
+			merge_input_config(current, ic);
+			current->input_type = device_type;
```

Comment:
This is just preference, but can you change the ordering to:
```c
current->input_type = device_type;
config->input_configs->items[i] = current;
free_input_config(ic);
```

It shouldn't make a difference, but it's a good idea to remove references before freeing.